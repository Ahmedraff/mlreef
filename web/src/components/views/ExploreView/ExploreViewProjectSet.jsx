import React, { useMemo } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import MProjectCard from 'components/ui/MProjectCard';
import MBricksWall from 'components/ui/MBricksWall';
import iconGrey from 'images/icon_grey-01.png';

const ExploreViewProjectSet = (props) => {
  const {
    started,
    projects,
    user: { id: userId },
    onMore,
  } = props;

  const loadingSection = useMemo(
    () => (
      <div className="loading-image">
        <div style={{ backgroundImage: 'url(/images/MLReef_loading.gif)' }} />
      </div>
    ),
    [],
  );

  const projectList = useMemo(
    () => (
      <>
        <div style={{ overflow: 'hidden' }}>
          <MBricksWall
            alignMarginLeft
            animated
            bricks={projects.map((proj) => (
              <MProjectCard
                className="bg-white"
                key={`proj-${proj.gitlabNamespace}-${proj.slug}`}
                slug={proj.slug}
                title={proj.name}
                projectId={proj.gitlabId}
                description={proj.description}
                starCount={proj.starCount || 0}
                forkCount={proj.forksCount || 0}
                namespace={proj.gitlabNamespace}
                updatedAt={proj.lastActivityat}
                projects={projects}
                dataProcessor={proj.dataProcessor}
                inputDataTypes={proj.inputDataTypes}
                outputDataTypes={proj.inputDataTypes}
                users={proj.members}
                visibility={proj.visibilityScope}
                owner={proj.ownerId === userId}
              />
            ))}
          />
        </div>
        {onMore && (
          <div className="d-flex">
            <button onClick={onMore} type="button" className="btn btn-link btn-hidden mx-auto py-3">
              More
            </button>
          </div>
        )}
      </>
    ),
    [projects, userId, onMore],
  );

  const checkEmpty = useMemo(
    () => projects.length > 0 ? projectList : (
      <div className="d-flex noelement-found-div">
        <img src={iconGrey} alt="" style={{ maxHeight: '100px' }} />
        <p>No projects found</p>
      </div>
    ),
    [projects, projectList],
  );

  return !started ? loadingSection : checkEmpty;
};

ExploreViewProjectSet.defaultProps = {
  projects: [],
  onMore: null,
};

ExploreViewProjectSet.propTypes = {
  started: PropTypes.bool.isRequired,
  projects: PropTypes.arrayOf(PropTypes.shape({})),
  onMore: PropTypes.func,
};

function mapStateToProps(state) {
  return {
    user: state.user,
  };
}

export default connect(mapStateToProps)(ExploreViewProjectSet);
