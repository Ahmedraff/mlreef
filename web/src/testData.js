export const imagesToRender = [
  {
    commitId: 'd80ab8ff15fb1f327bac0f5679b454cbec247759',
    parentCommitId: '6819ee9d2e8f7796887cc87af2d396c9894a2ebf',
    previousImage: {},
    nextImage: {},
    fileName: 'color_0_0054_A.png',
  },
];

export const mockMatchDataCommitDet = {
  path: '/my-projects/:projectId/commit/:commitId',
  url: '/my-projects/12395599/commit/aede036f526ea5d591f5ee56c131ed8472483912',
  isExact: true,
  params: {
    projectId: '12395599',
    branch: 'master',
    file: '.gitignore',
    commitId: 'aede036f526ea5d591f5ee56c131ed8472483912',
  },
};

export const commitMockObject = {
  id: 'd80ab8ff15fb1f327bac0f5679b454cbec247759',
  short_id: 'd80ab8ff',
  created_at: '2019-11-09T14:24:40.000Z',
  parent_ids: [
    '6819ee9d2e8f7796887cc87af2d396c9894a2ebf',
  ],
  title: 'Replace color_0_0054_A.png',
  message: 'Replace color_0_0054_A.png',
  author_name: 'cpmlreef',
  author_email: 'cp@mlreef.com',
  authored_date: '2019-11-09T14:24:40.000Z',
  committer_name: 'cpmlreef',
  committer_email: 'cp@mlreef.com',
  committed_date: '2019-11-09T14:24:40.000Z',
  stats: {
    additions: 0,
    deletions: 0,
    total: 0,
  },
  status: null,
  last_pipeline: null,
  project_id: 14448940,
};

export const mockDiffArray = [
  {
    old_path: 'showcase/color_0_0054_A.png',
    new_path: 'showcase/color_0_0054_A.png',
    a_mode: '100644',
    b_mode: '100644',
    new_file: false,
    renamed_file: false,
    deleted_file: false,
    diff: 'Binary files a/showcase/color_0_0054_A.png and b/showcase/color_0_0054_A.png differ\n',
  },
];

export const projectsArrayMock = {
  projects: {
    all: [
      {
        id: 14448940,
        description: 'Categorical Classification of alphabets in the American Sign Language',
        name: 'Sign Language Classifier',
        name_with_namespace: 'MLReef / Sign Language Classifier',
        path: 'sign-language-classifier',
        path_with_namespace: 'mlreef/sign-language-classifier',
        created_at: '2019-09-23T08:51:23.983Z',
        default_branch: 'master',
        tag_list: [],
        ssh_url_to_repo: 'git@gitlab.com:mlreef/sign-language-classifier.git',
        http_url_to_repo: 'https://gitlab.com/mlreef/sign-language-classifier.git',
        web_url: 'https://gitlab.com/mlreef/sign-language-classifier',
        readme_url: 'https://gitlab.com/mlreef/sign-language-classifier/blob/master/README.md',
        avatar_url: 'https://gitlab.com/uploads/-/system/project/avatar/14448940/proxy.duckduckgo.png',
        star_count: 0,
        forks_count: 0,
        last_activity_at: '2019-11-10T15:14:17.804Z',
        namespace: {
          id: 5351747,
          name: 'MLReef',
          path: 'mlreef',
          kind: 'group',
          full_path: 'mlreef',
          parent_id: null,
          avatar_url: '/uploads/-/system/group/avatar/5351747/MLReef_Logo_POS_H_icon-01.png',
          web_url: 'https://gitlab.com/groups/mlreef',
        },
      },
    ],
    selectedProject: {
      id: 12395599,
      description: 'This is MLReefÂ´s demo repository.',
      name: 'demo',
      name_with_namespace: 'MLReef / demo',
      path: 'mlreef-demo',
      path_with_namespace: 'mlreef/mlreef-demo',
      created_at: '2019-05-18T09:34:29.346Z',
      default_branch: 'master',
      tag_list: [],
      ssh_url_to_repo: 'git@gitlab.com:mlreef/mlreef-demo.git',
      http_url_to_repo: 'https://gitlab.com/mlreef/mlreef-demo.git',
      web_url: 'https://gitlab.com/mlreef/mlreef-demo',
      readme_url: 'https://gitlab.com/mlreef/mlreef-demo/blob/master/README.md',
      avatar_url: 'https://gitlab.com/uploads/-/system/project/avatar/12395599/shutterstock_129655604-Converted-copy-text.png',
      star_count: 2,
      forks_count: 1,
      last_activity_at: '2019-11-11T10:36:06.189Z',
      namespace: {
        id: 5351747,
        name: 'MLReef',
        path: 'mlreef',
        kind: 'group',
        full_path: 'mlreef',
        parent_id: null,
        avatar_url: '/uploads/-/system/group/avatar/5351747/MLReef_Logo_POS_H_icon-01.png',
        web_url: 'https://gitlab.com/groups/mlreef',
      },
    },
  },
  users: [
    {
      id: 3839940,
      name: 'cpmlreef',
      username: 'cpmlreef',
      state: 'active',
      avatar_url: 'https://assets.gitlab-static.net/uploads/-/system/user/avatar/3839940/avatar.png',
      web_url: 'https://gitlab.com/cpmlreef',
    },
  ],
};


export const branchesMock = [
  {
    name: 'data-instance/664e1c60',
    commit: {
      id: '12e05078d142b2bee947e86d6a52db4a5f85eb65',
      short_id: '12e05078',
      created_at: '2019-11-01T21:08:23.000+00:00',
      parent_ids: null,
      title: 'Add pipeline results [skip ci]',
      message: 'Add pipeline results [skip ci]',
      author_name: 'mlreefdemo',
      author_email: 'rainer+mlreefdemo@systemkern.com',
      authored_date: '2019-11-01T21:08:23.000+00:00',
      committer_name: 'mlreefdemo',
      committer_email: 'rainer+mlreefdemo@systemkern.com',
      committed_date: '2019-11-01T21:08:23.000+00:00',
    },
    merged: false,
    protected: false,
    developers_can_push: false,
    developers_can_merge: false,
    can_push: true,
    default: false,
  },
  {
    name: 'data-instance/9ae9c8c0',
    commit: {
      id: '3e36bfab31a1d43457e00f46af9e6d31d4f55a66',
      short_id: '3e36bfab',
      created_at: '2019-11-14T20:32:52.000+00:00',
      parent_ids: null,
      title: 'Add pipeline results [skip ci]',
      message: 'Add pipeline results [skip ci]',
      author_name: 'mlreefdemo',
      author_email: 'rainer+mlreefdemo@systemkern.com',
      authored_date: '2019-11-14T20:32:52.000+00:00',
      committer_name: 'mlreefdemo',
      committer_email: 'rainer+mlreefdemo@systemkern.com',
      committed_date: '2019-11-14T20:32:52.000+00:00',
    },
    merged: false,
    protected: false,
    developers_can_push: false,
    developers_can_merge: false,
    can_push: true,
    default: false,
  },
  {
    name: 'master',
    commit: {
      id: '3e36bfab31a1d43457e00f46af9e6d31d4f55a66',
      short_id: '3e36bfab',
      created_at: '2019-11-14T20:32:52.000+00:00',
      parent_ids: null,
      title: 'Add pipeline results [skip ci]',
      message: 'Add pipeline results [skip ci]',
      author_name: 'mlreefdemo',
      author_email: 'rainer+mlreefdemo@systemkern.com',
      authored_date: '2019-11-14T20:32:52.000+00:00',
      committer_name: 'mlreefdemo',
      committer_email: 'rainer+mlreefdemo@systemkern.com',
      committed_date: '2019-11-14T20:32:52.000+00:00',
    },
    merged: false,
    protected: true,
    developers_can_push: false,
    developers_can_merge: false,
    can_push: true,
    default: true,
  },
];
