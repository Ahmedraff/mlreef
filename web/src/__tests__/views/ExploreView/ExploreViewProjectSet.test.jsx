import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { projectsArrayMock } from 'testData';
import ExploreViewProjectSet from 'components/views/ExploreView/ExploreViewProjectSet';

const projects = projectsArrayMock.projects.all;

describe('ExploreViewProjectSet functionality', () => {
  global.ResizeObserver = () => ({ observe: jest.fn() });
  test('assert only loading gif is showed while starting', () => {
    const wrapper = mount(
      <ExploreViewProjectSet started={false} projects={projects} />,
    );

    expect(wrapper.find('MProjectCard')).toHaveLength(0);
    expect(wrapper.find('.loading-image')).toHaveLength(1);
    expect(wrapper.find('.noelement-found-div')).toHaveLength(0);
  });

  test('assert only empty icon is showed for no project', () => {
    const wrapper = mount(
      <ExploreViewProjectSet started projects={[]} />,
    );

    expect(wrapper.find('MProjectCard')).toHaveLength(0);
    expect(wrapper.find('.loading-image')).toHaveLength(0);
    expect(wrapper.find('.noelement-found-div')).toHaveLength(1);
  });

  test('assert MProjectCard is showed', () => {
    const wrapper = mount(
      <ExploreViewProjectSet started projects={projects} />,
    );

    expect(wrapper.find('MProjectCard')).toHaveLength(1);
    expect(wrapper.find('.loading-image')).toHaveLength(0);
    expect(wrapper.find('.noelement-found-div')).toHaveLength(0);
  });
});

describe('ExploreViewProjectSet appareance', () => {
  test('assert that snapshot matches', () => {
    const snapshot = renderer.create(<ExploreViewProjectSet started projects={projects} />)
      .toJSON();

    expect(snapshot).toMatchSnapshot();
  });
});
