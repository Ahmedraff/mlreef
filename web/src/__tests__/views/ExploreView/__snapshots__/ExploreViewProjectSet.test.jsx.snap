// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExploreViewProjectSet appareance assert that snapshot matches 1`] = `
<div
  style={
    Object {
      "overflow": "hidden",
    }
  }
>
  <div
    className="m-bricks-wall"
  >
    <div
      className="m-bricks-wall-brick animated"
    >
      <div
        className="card"
      >
        <img
          alt="private"
          aria-hidden="true"
          className="icon-visibility"
          height="24"
          src="/images/Lock-01.svg"
          width="24"
        />
        <div
          className="card-container project-card-container"
        >
          <a
            className="project-card-link"
            href="/Some namespace/some-slug"
            onClick={[Function]}
          >
            <p
              className="card-title"
            >
              Sign Language Classifier
            </p>
            <div
              className="project-card-container-autor pb-3 mt-2"
            >
              <p>
                Some namespace
              </p>
            </div>
            <div
              className="card-content"
            >
              <div
                className="project-card-types d-flex"
              />
              <section
                className="m-paragraph card-content-description"
              >
                <p
                  className="m-paragraph-line "
                >
                  Categorical Classification of alphabets in the American Sign Language
                </p>
              </section>
              <div
                className="d-flex t-secondary"
              >
                <div
                  className="mr-3"
                >
                  <i
                    className="fa fa-star"
                  >
                    <span
                      className="label"
                    >
                       0
                    </span>
                  </i>
                </div>
                <div
                  className="mr-3"
                >
                  <i
                    className="fa fa-code-branch"
                  >
                    <span
                      className="label"
                    >
                       0
                    </span>
                  </i>
                </div>
              </div>
            </div>
          </a>
          <div
            className="card-actions project-card-actions"
          >
            <div
              className="avatars-reversed"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
